### ISSUE TITLE
Slow startup due to deadlock if o.fd.FileManager1 is implemented by GtkApplication

### ISSUE THREAD

**Issue by minhduc0711** [flatpak/xdg-desktop-portal#636]
As a result, several apps have to wait while the service was starting.

This starts happening when I updated to 1.10.1. Notice the time between the `Starting Portal service` and `Started Portal service`:
```
● xdg-desktop-portal.service - Portal service
     Loaded: loaded (/usr/lib/systemd/user/xdg-desktop-portal.service; static)
     Active: active (running) since Fri 2021-09-24 10:24:12 +07; 46s ago
   Main PID: 31907 (xdg-desktop-por)
      Tasks: 10 (limit: 23842)
     Memory: 3.7M
        CPU: 392ms
     CGroup: /user.slice/user-1000.slice/user@1000.service/app.slice/xdg-desktop-portal.service
             └─31907 /usr/lib/xdg-desktop-portal

Sep 24 10:23:47 acer-aspire-a5 systemd[924]: Starting Portal service...
Sep 24 10:24:12 acer-aspire-a5 systemd[924]: Started Portal service.
Sep 24 10:24:42 acer-aspire-a5 xdg-desktop-por[31907]: Failed to get application states: GDBus.Error:org.freedesktop.portal.Error.Failed: Could not get window list
```

The service started immediately in 1.8.1:
```
● xdg-desktop-portal.service - Portal service
     Loaded: loaded (/usr/lib/systemd/user/xdg-desktop-portal.service; static)
     Active: active (running) since Fri 2021-09-24 10:26:46 +07; 1s ago
   Main PID: 34405 (xdg-desktop-por)
      Tasks: 8 (limit: 23842)
     Memory: 3.1M
        CPU: 101ms
     CGroup: /user.slice/user-1000.slice/user@1000.service/app.slice/xdg-desktop-portal.service
             └─34405 /usr/lib/xdg-desktop-portal

Sep 24 10:26:46 acer-aspire-a5 systemd[924]: Starting Portal service...
Sep 24 10:26:46 acer-aspire-a5 systemd[924]: Started Portal service.                                                    
```

FYI I also use sway, so there might be some configuration problems.
Edit: I checked https://github.com/flatpak/xdg-desktop-portal/issues/607, but my xdg-desktop-portal-wlr seems to start up properly.

**Comment by AidanGG** [comment]
Do you happen to use `xdg-desktop-portal-gtk` and some GTK-based file manager? Perhaps the Reddit thread here might be similar to your issue: https://www.reddit.com/r/swaywm/comments/ptwd6p/thunar_and_xdgdesktopportal_block_each_other/

**Comment by minhduc0711** [comment]
Thank you, I removed `xdg-desktop-portal-gtk` and the issue is gone. I will just gonna leave this issue open for now in case it's a bug in xdg-desktop-portal.

**Comment by smcv** [comment]
> Do you happen to use xdg-desktop-portal-gtk and some GTK-based file manager?

This issue was also reported in Debian, and from debug information provided by the bug reporter, this is the key factor.

The problem is that xdg-desktop-portal calls into org.freedesktop.FileManager1 on startup (to find out whether it can be used to implement the OpenURI interface), and GTK's GtkApplication class (as used in e.g. Thunar) calls into the Inhibit portal on startup (to use it as a desktop-neutral way to offer an interface to inhinit screensaver/suspend/shutdown). This would be fine if they both used asynchronous D-Bus calls and continued to start up without waiting for a result, but they don't: they both use synchronous (blocking) D-Bus calls. This means neither one will successfully start up until it gets a reply from the other, causing a deadlock. The dbus-daemon waits for an arbitrary length of time for the two services to start up, then resolves the deadlock by deciding that service-activation is taking too long and sending an error reply.

This will only happen in non-GNOME environments. If gnome-session is running, GTK bypasses the Inhibit portal abstraction and uses gnome-session's D-Bus interfaces directly, so this will not occur.

> I removed xdg-desktop-portal-gtk and the issue is gone. I will just gonna leave this issue open for now in case it's a bug in xdg-desktop-portal.

It's a bug in xdg-desktop-portal (and/or GTK, but I think it's going to be a lot easier to fix here than in GTK). Removing the -gtk backend probably only works around this by coincidence, because the -gtk backend provides the AppChooser interface (which is required for OpenURI) and the -wlr backend does not.

xdg-desktop-portal will not have full functionality without an implementation of the AppChooser interface.

**Comment by smcv** [comment]
This appears to have regressed in #601, while fixing #564.

**Comment by smcv** [comment]
> This will only happen in non-GNOME environments

After wasting some time trying to reproduce this in XFCE, make that "non-GNOME && non-XFCE environments"... (GTK has specific support for XFCE's session manager, too.)

**Comment by pietryszak** [comment]
> Any chance to push out a new release including this fix?

Why new version with this fix is not released yet ?

**Comment by WhyNotHugo** [comment]
Can a new release with this fix be pushed out? This adds an artificial delay at system startup that's very annoying.

**Comment by Mahrjose** [comment]
Is this fixed yet? I'm having similar problem. Installing flatpak package in arch linux (xmonad) slows down thunar startup time by 25-30s! Without flatpak installed startup time is only 2-3s.

**Comment by hyuri** [comment]
Still happening. Chromium takes 5 minutes to open after startup.

**Comment by smcv** [comment]
#639 was included in 1.12.0 and all newer versions.

@Mahrjose, @hyuri, you might be experiencing the same issue as #986 and #1024, rather than what was originally reported here. If so, xdg-desktop-portal 1.17.0 should resolve that.

If you are experiencing a similar symptom with xdg-desktop-portal 1.17.0 or later, please open a separate issue with more information, such as the versions of relevant components that you are using (xdg-desktop-portal, any backends such as xdg-desktop-portal-gtk that you have installed) and the messages that appear in the systemd journal during the delay.
